src/widget/switch.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A toggle switch widget.
+
+use crate::kurbo::{Circle, Point, Rect, RoundedRect, Size};
+use crate::piet::{FontBuilder, Text, TextLayout, TextLayoutBuilder};
+use crate::piet::{LinearGradient, RenderContext, UnitPoint};
+use crate::theme;
+use crate::widget::Align;
+use crate::{
+    BaseState, BoxConstraints, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+#[derive(Debug, Clone)]
+pub struct Switch;
+
+impl Switch {
+    pub fn new() -> impl Widget<bool> {
+        Align::vertical(UnitPoint::CENTER, SwitchRaw::default())
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SwitchRaw {
```

Comment:
I am not a huge fan of this 'Widget + WidgetRaw' pattern; I don't think it should be the default pick for most widgets.

In this case, your `Switch` statement doesn't have any state; it's really just being used as a marker, and as somewhere to hang the `new` method. I would delete SwitchRaw` and just put that state in `Switch` itself, and make it the widget.